src/compiled/version/Token.hpp

```diff
@@ -0,0 +1,245 @@
+#ifndef REZ_TOKEN_HPP
+#define REZ_TOKEN_HPP
+
+#include "Comparable.hpp"
+
+#include <regex>
+
+template<typename _Typ> struct TokenData
+{
+    using value_type = _Typ;
+
+    // data access
+    const value_type& Get() const REZ_NOEXCEPT { return _value; }
+    value_type& Get()  REZ_NOEXCEPT { return _value; }
+
+    // operator->
+    const value_type* operator->() const REZ_NOEXCEPT { return &Get; }
+    value_type* operator->() REZ_NOEXCEPT { return &Get(); }
+
+    // operator*
+    const value_type& operator*() const REZ_NOEXCEPT { return Get(); }
+    value_type& operator*() REZ_NOEXCEPT { return Get(); }
+
+    // operator<
+    bool operator<(const TokenData<value_type>& other) const REZ_NOEXCEPT
+    {
+        return _value < other.Get();
+    }
+
+    // operator<
+    bool operator<(const value_type& other) const REZ_NOEXCEPT
+    {
+        return _value < other;
+    }
+
+    // operator==
+    bool operator==(const TokenData<value_type>& other) const REZ_NOEXCEPT
+    {
+        return _value == other.Get();
+    }
+
+    // operator==
+    bool operator==(const value_type& other) const REZ_NOEXCEPT
+    {
+        return _value == other;
+    }
+
+    value_type _value;
+};
+
+template<typename T> class SubToken
+{
+public:
+    explicit SubToken(const T& v)
+    {
+        s = v;
+        n = is_digit(s) ? to_int(s) : REZ_INT_INVALID;
+    }
+
+    explicit SubToken(T&& v)
+    {
+        s = std::move(v);
+        n = is_digit(s) ? to_int(s) : REZ_INT_INVALID;
+    }
+
+    // operator<
+    bool operator<(const SubToken& other) const REZ_NOEXCEPT
+    {
+        if (n == REZ_INT_INVALID)
+        {
+            return other.n == REZ_INT_INVALID ? s < other.s : true;
+        }
+
+        return other.n == REZ_INT_INVALID ? false : std::tie(n, s) < std::tie(other.n, other.s);
+    }
+
+    // operator<
+    bool operator<(rez_int other) const REZ_NOEXCEPT
+    {
+        return other == REZ_INT_INVALID ? false : n < other;
+    }
+
+    // operator==
+    bool operator==(const SubToken& other) const REZ_NOEXCEPT
+    {
+        return (s == other.s) && (n == other.n);
+    }
+
+    // operator==
+    bool operator==(rez_int other) const REZ_NOEXCEPT
+    {
+        return other == REZ_INT_INVALID ? false : n == other;
+    }
+
+    T s;
+    rez_int n;
+};
+
+// is_token type
+template<typename _Typ> struct is_token : std::false_type
+{
+};
+
+template<typename _Typ, bool _Rev> struct is_token<Comparable<TokenData<_Typ>, _Rev>> : std::true_type
+{
+};
+
+//
+// Numeric Token
+//
+using NumericValue = rez_int;
+template<bool _Rev> using NumericTokenT = Comparable<TokenData<NumericValue>, _Rev>;
+
+using NumericToken = NumericTokenT<NORMAL>;
```

Comment:
`NumericToken` can be replaced by `rez_int` and factory specialization to create `NumericVersion`.